server/clip_server/model/clip.py

```diff
@@ -81,53 +95,67 @@ def _download(url: str, root: str, with_resume: bool = True):
-
-                progress.start_task(task)
-
-                if resume_byte_pos and with_resume:
-                    progress.update(task, advance=resume_byte_pos)
-                    req.headers['Range'] = f'bytes={resume_byte_pos}-'
-
-                with urllib.request.urlopen(req) as source:
-                    while True:
-                        buffer = source.read(8192)
-                        if not buffer:
-                            break
-
-                        output.write(buffer)
-                        progress.update(task, advance=len(buffer))
-        except Exception as ex:
-            raise ex
-        finally:
-            # rename the temp download file to the correct name if fully downloaded
-            if os.path.exists(tmp_file_path) and (
-                total_bytes == os.path.getsize(tmp_file_path)
-            ):
-                shutil.move(tmp_file_path, download_target)
-
-    return download_target
+            total_bytes = -1
+            try:
+                # resolve the 403 error by passing a valid user-agent
+                req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
+                total_bytes = int(
+                    urllib.request.urlopen(req).info().get('Content-Length', -1)
+                )
+                mode = 'ab' if (with_resume and resume_byte_pos) else 'wb'
+
+                with open(tmp_file_path, mode) as output:
+                    progress.update(task, total=total_bytes)
+                    progress.start_task(task)
+
+                    if resume_byte_pos and with_resume:
+                        progress.update(task, advance=resume_byte_pos)
+                        req.headers['Range'] = f'bytes={resume_byte_pos}-'
+
+                    with urllib.request.urlopen(req) as source:
+                        while True:
+                            buffer = source.read(8192)
+                            if not buffer:
+                                break
+
+                            output.write(buffer)
+                            progress.update(task, advance=len(buffer))
+            except Exception as ex:
+                progress.console.print(f'Error: {ex} Retrying now...')
+                continue
+
+            finally:
+                # rename the temp download file to the correct name if fully downloaded
+                if os.path.exists(tmp_file_path) and (
+                    total_bytes == os.path.getsize(tmp_file_path)
+                    and (
+                        not md5
+                        or hashlib.md5(open(tmp_file_path, 'rb').read()).hexdigest()
+                        == md5
+                    )
+                ):
+                    shutil.move(tmp_file_path, download_target)
+                    break
+                else:
+                    progress.console.print(
+                        f'MD5 mismatch for {download_target}, maybe the file is not completely downloaded. '
```

Comment:
can we output the expected and real md5 hash code? This will help us to debug when the user has trouble downloading the model.